### DISCUSSION TITLE
Error: AppId changed without project login

### DISCUSSION THREAD

**Discussion by Lex-Kent** [speechly/speechly#259] [General]
Hi all,

Here I am again, getting absolutely nowhere with the Speechly documentation!

I am trying to switch Applications in order to target different intents for specific user responses within my app. As usual I am struggling!

I successfully stopped the `Speechly.BrowserClient()` and recreated a new one with a different AppID but this slowed the responsiveness of my web page enormously. So now I am trying to simply `stop()` and then `start({appId : someDifferentAppID})`, but when I do I get the error `Error: AppId changed without project login` in my console.

What am I doing wrong please?

**Comment by teelisyys** [comment]
Hi @Lex-Kent !

If the BrowserClient is originally created with an _app id_, then the app cannot be changed at the `start`, as the login was only made with the original app id.

However, the BrowserClient supports _project login_ where the BrowserClient is created with a [_project id_](https://github.com/speechly/speechly/blob/main/libraries/browser-client/docs/interfaces/client.DecoderOptions.md#projectid), then you can switch between apps in that project.

~~~
const client = new BrowserClient({ projectId: 'your-project-id' });
~~~

You can find the project id at "Project settings" in the Dashboard.

**Comment by Lex-Kent** [comment]
@teelisyys thank you for your very quick response, as usual!

I'm now using the Project ID like so:

```
ctx.speechly = new BrowserClient({
  projectId: "myProjectID",
  logSegments: false,
  vad: {
    enabled: true,
    noiseGateDb: -24.0
  }
});
```
and sometimes it works, but half the time, on page load, it fails with this error:

![image](https://user-images.githubusercontent.com/107930534/208137055-a79e8e93-d24c-42ca-9832-98fe3c126a8c.png)

And if I add the appID:

```
ctx.speechly = new BrowserClient({
  projectId: "myProjectID",
  appId: ctx.speechlyAppID,
  logSegments: false,
  vad: {
    enabled: true,
    noiseGateDb: -24.0
  }
});
```

then I get this error instead :

![image](https://user-images.githubusercontent.com/107930534/208137531-a1614e36-9214-486c-bc8d-9b1c72c6aa5e.png)

In other circumstances it might be funny !

**Comment by Lex-Kent** [comment]
@teelisyys I think I've figured it out.

Enabling `vad` in my options meant that the `BrowserClient` would start automatically if I made a sound above a certain threshold. Calling `start` in this fashion, without the `appID` caused the error. I'm assuming the reason it was intermittent is because sometimes when I loaded the page, I didn't make any noise loud enough to invoke the `vad`.

So I've removed `vad` from the initial `BrowserClient` creation and it now works as expected.

Thanks again for you help! :o)

**Comment by Lex-Kent** [comment]
I'm still not quite there with this!

I've successfully created my `BrowserClient` with a project ID, and then I call `start()` with an appID to point it to the correct Application for that particular user response.

My "default" Application has no training data, it's the Speechly default. When I get results back from this Application, my `segment` has no `intent` or `entities` - I use my own code to interrogate the words. My other Application has training data and _does_ return an `intent` and `entities`.

My code is failing every 3rd or 4th time I execute it because my `segment` which should include an `intent` and `entities` does not. It looks as though my `BrowserClient` is still using my default Application despite me calling `await MyBrowserClient.stop()` and then `MyBrowserClient.start({appId: myOtherAppID})` to switch Applications.

I can't tell for sure, however because trying to log out `MyBrowserClient.decoder.appId` in the `onSegmentChange` event returns `undefined`.

Can anyone help please? I'm so close to having this working!

**Comment by Lex-Kent** [comment]
Okay, I think I can see what's happening. My response was coming back with no `intent` or `entities` when I was giving a response that Speechly couldn't match to my training data.

I've now created a separate app for each question and now I get an `intent` back every time, and if I submit a transcript that cannot be matched, then my `entities` array is empty, but at least I know I was targeting the correct application!
